[MRG] Fix #6031: changed calculation of explained_variance_ratio_, SVD solver

JPFrancoia opened the pull request:
solver. See issue #5216. However, this attribute is corrected from the
previous commit which might not return the expected values.

Commit:
Attribute explained_variance ratio is now available for LDA with the SVD
solver. See issue #5216. However, this attribute is corrected from the
previous commit which, might not return the expected values.

agramfort commented:
this is an API change. You cannot do this. The size of the attribute must stay the same. Otherwise we need a deprecation cycle.

JPFrancoia commented:
Yeah, but we have to anyway. self.explained_variance_ratio_ should return an array of length n if you give the LDA class n dimensions. It is not consistent with:

https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/discriminant_analysis.py#337

And it is also not consistent with the same attribute of the PCA class. See #5216.

This should not have been merged. It should at least be:

```python
self.explained_variance_ratio_ = S / S.sum()
```

And only if I'm wrong in #5216

Commit:
Corrected the dimension of the array explained_variance_ratio_ for LDA,
solver svd.

JPFrancoia commented:
Ok. Then https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/discriminant_analysis.py#L337 is not consistent, and should be:

```python
self.explained_variance_ratio_ = np.sort(evals / np.sum(evals))[::-1][:self.n_components]
```

agramfort commented:
ok please update the unit tests to show this issue and present the PR as a
fix.

thx

hlin117 commented:
@JPFrancoia: Like @agramfort mentioned, you should change the title of the PR to allude to a bug fix. Thanks!

JPFrancoia commented:
Like that ?

hlin117 commented:
@JPFrancoia : Excellent =]

Good luck!

JPFrancoia commented:
I'll make the corrections asap.

Commit:
Cleaned the calculation of explained_variance_ratio_ in
LinearDiscriminantAnalysis, SVD solver.

Commit:
Regression test for explained_variance_ratio_, LDA, SVD solver.

Used the script of @hlin117 in #5216 to generate better test data,
make_blobs was not good enough.

JPFrancoia commented:
@hlin117 , for the regression tests, I used your way to generate the test data, it's efficient.

I think it generates better test data than make_blobs, which typically returns an array like:

[1 a_very_small_value]

With this kind of array, the test would always pass, and wouldn't spot errors. Now it will.

hlin117 commented on sklearn/discriminant_analysis.py:
```diff
@@ -397,7 +397,8 @@ def _solve_svd(self, X, y):
 
-        self.explained_variance_ratio_ = S[:self.n_components] / S.sum()
+        self.explained_variance_ratio_ = (S**2 / np.sum(
+            S**2))[:self.n_components]
```
Add 4 more spaces here to conform to PEP 8 indentation.
https://www.python.org/dev/peps/pep-0008/#indentation

hlin117 commented on sklearn/tests/test_discriminant_analysis.py:
```diff
@@ -176,8 +175,12 @@ def test_lda_explained_variance_ratio():
+    # before comparison.
     assert_array_almost_equal(clf_lda_svd.explained_variance_ratio_,
-                              clf_lda_eigen.explained_variance_ratio_)
+            clf_lda_eigen.explained_variance_ratio_[:3])
```
That's a good line note. But why aren't you passing in `n_components=3` into line 171 and line 175?

In my opinion, you should just compare both of the arrays without the slice. They should theoretically be of the same length anyways.

hlin117 replied:
@JPFrancoia: Sorry, your'e right. They're not of the same length.

I think what would be more transparent than the 3 is using `tested_length = min(clf_lda_svd.explained_variance_ratio_.shape[0], clf_lda_eigen.explained_variance_ratio_.shape[0])`.

The extra length is caused by precision errors in this case, not because of the issue mentioned in #6032. #6032 deals with the fact when `n_components` is set in the constructor, but this length is not reflected by `explained_variance_ratio_`. On the other hand, in this case, we are not setting `n_components` in either constructor, but we're getting different lengths due to precision errors.

JPFrancoia replied:
True. What do you mean by "precision errors" ?

hlin117 replied:
I'd accept that length, if it were me.

By precision errors, I mean that the arrays you showed before

```
eigen
20
[  6.03795532e-01   3.96204468e-01   4.19383626e-16   3.08212423e-16
   1.23275796e-16   9.84308055e-17   6.96952083e-17   4.57224815e-17
   3.73609416e-17   2.12787319e-17   9.54458789e-18  -1.83915898e-17
  -4.11632284e-17  -1.08854065e-16  -1.41296422e-16  -1.58466942e-16
  -1.87047513e-16  -2.18274226e-16  -3.00368948e-16  -4.68795358e-16]
svd
3
[  6.03795532e-01   3.96204468e-01   1.23375797e-32]
Traceback (most recent call last):
  File "/home/djipey/informatique/python/scikit-learn/sklearn/tests/test_discriminant_analysis.py", line 368, in <module>
    test_lda_explained_variance_ratio()
  File "/home/djipey/informatique/python/scikit-learn/sklearn/tests/test_discriminant_analysis.py", line 193, in test_lda_explained_variance_ratio
    clf_lda_eigen.explained_variance_ratio_)
  File "/usr/lib/python3.5/site-packages/numpy/testing/utils.py", line 886, in assert_array_almost_equal
    precision=decimal)
  File "/usr/lib/python3.5/site-packages/numpy/testing/utils.py", line 663, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 6 decimals
```

seem very solver dependent. The eigen solver might be more precise, and grabbing the eigenvalues that are extremely close to zero. 

Related topic, numerical stability:
https://en.wikipedia.org/wiki/Numerical_stability

JPFrancoia replied:
Yes but why is the first array of length 20, and the second of length 3 ?

JPFrancoia replied:
I mean, why is the sdv array of length 3 ? We don't specify n_components when we initialize the classifiers.

hlin117 replied:
If `n_components=None`, then by default, the svd and eigen
LinearDiscriminantAnalysis classes will return all of the eigenvalues. It
just turns out that the eigen solver yields more eigenvalues than the SVD
solver.

On Tue, Dec 15, 2015 at 5:29 PM, JPFrancoia <notifications@github.com>
wrote:

> In sklearn/tests/test_discriminant_analysis.py
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_scikit-2Dlearn_scikit-2Dlearn_pull_6027-23discussion-5Fr47712339&d=BQMCaQ&c=8hUWFZcy2Z-Za5rBPlktOQ&r=IgVP_sCIQHhgRgxQoo7gufqSQOvlFIckTGgYIpXBa_4&m=AY2fz0E96atHTtLqc2ACdb4QVEt00CicRvw0s637MTI&s=Zu0YSrErMT2_MEUBieaV6Oxmv6w7p4Qp3yVNkt1lUCI&e=>
> :
>
> >      assert_array_almost_equal(clf_lda_svd.explained_variance_ratio_,
> > -                              clf_lda_eigen.explained_variance_ratio_)
> > +            clf_lda_eigen.explained_variance_ratio_[:3])
>
> I mean, why is the sdv array of length 3 ? We don't specify n_components
> when we initialize the classifiers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_scikit-2Dlearn_scikit-2Dlearn_pull_6027_files-23r47712339&d=BQMCaQ&c=8hUWFZcy2Z-Za5rBPlktOQ&r=IgVP_sCIQHhgRgxQoo7gufqSQOvlFIckTGgYIpXBa_4&m=AY2fz0E96atHTtLqc2ACdb4QVEt00CicRvw0s637MTI&s=GQ8PKCIJDMszV5ezTabub8KKiTGyOE7H6dg4rMfFm08&e=>
> .
>



-- 



*Henry Lin, Research AssistantM.S. Computer ScienceUniversity of Illinois
at Urbana-Champaign 2016847-769-8729 | halin2@illinois.edu
<halin2@illinois.edu>*

JPFrancoia commented:
> They should theoretically be of the same length anyways.

They are not:

```python
def test_lda_explained_variance_ratio():
    # Test if the sum of the normalized eigen vectors values equals 1,
    # Also tests whether the explained_variance_ratio_ formed by the
    # eigen solver is the same as the explained_variance_ratio_ formed
    # by the svd solver

    state = np.random.RandomState(0)
    X = state.normal(loc=0, scale=100, size=(40, 20))
    y = state.randint(0, 3, size=(40, 1))

    clf_lda_eigen = LinearDiscriminantAnalysis(solver="eigen")
    clf_lda_eigen.fit(X, y)
    assert_almost_equal(clf_lda_eigen.explained_variance_ratio_.sum(), 1.0, 3)

    print("eigen")
    print(len(clf_lda_eigen.explained_variance_ratio_))
    print(clf_lda_eigen.explained_variance_ratio_)

    clf_lda_svd = LinearDiscriminantAnalysis(solver="svd")
    clf_lda_svd.fit(X, y)
    assert_almost_equal(clf_lda_svd.explained_variance_ratio_.sum(), 1.0, 3)

    print("svd")
    print(len(clf_lda_svd.explained_variance_ratio_))
    print(clf_lda_svd.explained_variance_ratio_)

    # NOTE: clf_lda_eigen.explained_variance_ratio_ is not of n_components
    # length. Make it the same length as clf_lda_svd.explained_variance_ratio_
    # before comparison.
    assert_array_almost_equal(clf_lda_svd.explained_variance_ratio_,
            clf_lda_eigen.explained_variance_ratio_)
```


```
eigen
20
[  6.03795532e-01   3.96204468e-01   4.19383626e-16   3.08212423e-16
   1.23275796e-16   9.84308055e-17   6.96952083e-17   4.57224815e-17
   3.73609416e-17   2.12787319e-17   9.54458789e-18  -1.83915898e-17
  -4.11632284e-17  -1.08854065e-16  -1.41296422e-16  -1.58466942e-16
  -1.87047513e-16  -2.18274226e-16  -3.00368948e-16  -4.68795358e-16]
/usr/lib/python3.5/site-packages/sklearn/utils/validation.py:515: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
svd
3
[  6.03795532e-01   3.96204468e-01   1.23375797e-32]
Traceback (most recent call last):
  File "/home/djipey/informatique/python/scikit-learn/sklearn/tests/test_discriminant_analysis.py", line 368, in <module>
    test_lda_explained_variance_ratio()
  File "/home/djipey/informatique/python/scikit-learn/sklearn/tests/test_discriminant_analysis.py", line 193, in test_lda_explained_variance_ratio
    clf_lda_eigen.explained_variance_ratio_)
  File "/usr/lib/python3.5/site-packages/numpy/testing/utils.py", line 886, in assert_array_almost_equal
    precision=decimal)
  File "/usr/lib/python3.5/site-packages/numpy/testing/utils.py", line 663, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 6 decimals
```

JPFrancoia committed:
Added 4 spaces to conform to PEP8.

JPFrancoia commented:
Of course they are not. See #6032

JPFrancoia committed:
Modified the length of the slice: now uses a variable, not a hardcoded
number.

JPFrancoia commented:
Ok, corrections made @hlin117 .

hlin117 commented:
LGTM. Can you add [MRG] to the title of this PR?

JPFrancoia commented:
Done. I think we can close #6031 and #5216 when this PR is merged.

hlin117 commented:
@JPFrancoia : #5216 has already been merged. Did you mean another PR or issue?

JPFrancoia commented:
Hi, I'm actually not sure what the [MRG] tag means. Will this PR be merged one day ? Do I need to do something more ?

hlin117 commented:
@JPFrancoia [MRG] means that it is ready to be merged, and it is until review from the scikit-learn team.

agramfort commented:
LGTM @JPFrancoia please update the what's new to document the bug fix and we're good

JPFrancoia committed:
Modified the What's new file to mention the bug fix about the wrong
results returned with discriminant_analysis.LinearDiscriminantAnalysis's
attribute explained_variance_ratio.

JPFrancoia commented:
Crap. One test failed, and there are some conflicts. What is AppVeyor ? I don't know it.

agramfort commented:
merged by rebase after a couple of fixes 

0476784cb0ddab5389234cb9479e4ea3a4e0cfc6
74c29e0bd06c0812435615b1d7c00fbec1593d48
262e9e476a84b4aca00199d2cdbbe1956684a6d3

thx @JPFrancoia
